code/modules/power/debug/power_debugger.dm

```diff
@@ -0,0 +1,322 @@
+			"pw_channel" = local_net.channel_to_name(machine.power_channel), //we love helper procs!
+			"pw_state" = local_net.state_to_name(machine.power_state),
+			"idle_consumption" = machine.idle_power_consumption,
+			"active_consumption" = machine.active_power_consumption,
+			"type" = machine.type,
+			"dir" = dir2text(machine.dir),
+		)
+		powernet_data["local_machines"] += list(machine_data)
+		get_power_icon(machine, machine.dir)
+	powernet_data["logs"] = selected_local_net.powernet_log
+	return powernet_data
+
+
+/datum/ui_module/powernet_debugger/ui_data(mob/user)
+	var/list/data = list()
+
+	// # Filter Data
+	data["filters"] = list(
+		"pipes" = filter_pipes,
+		"lights" = filter_lights,
+		"apcs" = filter_apcs,
+		"terminals" = filter_terminals,
+		"off_station" = filter_non_station_powernets
+	)
+
+	// Global info
+	data["debug_page"] = debug_screen
+	// all powernets we have actively selected, this generates the tabs in our UI
+	data["selected_nets"] = selected_powernets
+	// list of all regional powernets
+	data["powernets"] = get_all_powernets()
+	// Detailed page info
+	data["selected_net"] = get_selected_powernet_data()
+	// a copy of our power icon cache, used to generate pictures of power machines for use in our UI's tables
+	data["power_images"] = power_icon_cache
+
+	return data
+
+/// Since we need icons for our UI, this proc will generate the icon if and only if the icon for that type in the specified direction does not already
+/// have its icon generated. Otherwise it pulls the icon in that dir from the cache. Avoids having to regenerate icons every data update!
+/datum/ui_module/powernet_debugger/proc/get_power_icon(obj/power_machine, machine_dir = SOUTH)
+	if(!machine_dir) // need a direction or this proc runtimes, default to south if null
+		machine_dir = SOUTH
+	if(!power_icon_cache[power_machine.type]) // have we cached this type yet?
+		power_icon_cache[power_machine.type] = list()
+	if(!power_icon_cache[power_machine.type][dir2text(machine_dir)]) // have we cached this type in the specified direction yet?
+		power_icon_cache[power_machine.type][dir2text(machine_dir)] = "[icon2base64(icon(power_machine.icon, power_machine.icon_state, machine_dir))]"
+	return power_icon_cache[power_machine.type][dir2text(machine_dir)] // return the icon incase we need it!
+
+/datum/ui_module/powernet_debugger/ui_act(action, params, datum/tgui/ui)
+	if(!check_rights(R_DEBUG))
+		return
+	if(..())
+		return
+
+	switch(action)
+		if("detailed_view")
+			var/datum/powernet = locateUID(params["PW_UID"])
+			if(istype(powernet, /datum/regional_powernet))
+				selected_regional_net = powernet
+				selected_local_net = null
+			else if(istype(powernet, /datum/local_powernet))
+				selected_local_net = powernet
+				selected_regional_net = null
+			else
+				return
+			selected_powernets |= params["PW_UID"]
+			debug_screen = PW_DEBUG_SCREEN_DETAILS
+		if("open_vv")
+			var/client/C = ui.user.client
+			if(!check_rights(R_ADMIN))
```

Comment:
To be safe
```suggestion
			if(!check_rights(R_ADMIN), ui.user)
```

Reply:
On another note, this would probably be okay to move to developer rights as well (and helpful).